### PR TITLE
SAAS-3765: Added SNS Appropriate Subscribers Plugin

### PR DESCRIPTION
**Pull request by sadeed12345**

### CODE REVIEW

File: plugins/aws/sns/snsAppropriateSubscribers.js [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    more_info: 'Amazon Simple Notification Service (Amazon SNS) is a managed service that provides message delivery from publishers to subscribers. So check for appropriate subsribers in order to improve access security to your SNS topics. ',
+    recommended_action: 'Ensure that no unwanted subscribers are using the subscriptions periodically.',
+    link: 'https://docs.aws.amazon.com/sns/latest/dg/sns-create-subscribe-endpoint-to-topic.html',
+    apis: ['SNS:listSubscriptions'],
+    settings: {
+        sns_unwanted_subscribers_endpoint: {
+            name: 'SNS Unwanted Subscribers Endpoint',
+            description: 'endpoint for SNS subscrptions that are unwanted',
+            regex: '^.*$',
+            default: 'sadeed1999@gmail.com',
```
Comment (sadeed12345) [inline plugins/aws/sns/snsAppropriateSubscribers.js:18]:
```suggestion
            default: '',
```

File: plugins/aws/sns/snsAppropriateSubscribers.js [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+                let resource = subscriber.SubscriptionArn;
+
+                if (subscriber.Endpoint == config.sns_unwanted_subscribers_endpoint){
+                    helpers.addResult(results, 2,
+                        'SNS subscriber is unwanted for topic', region, resource);
+                } else {
+                    helpers.addResult(results, 0,
+                        'SNS subscriber is appropriate for topic', region, resource);
+                }
+
```
Comment (sadeed12345) [inline plugins/aws/sns/snsAppropriateSubscribers.js:61]:
```suggestion
```

File: collectors/aws/collector.js [hunk 3]
```diff
@@ -1767,7 +1771,7 @@ var postcalls = [
                 reliesOnCall: 'listTopics',
                 filterKey: 'TopicArn',
                 filterValue: 'TopicArn'
-            }
+            },
```
Comment (AkhtarAmir) [inline collectors/aws/collector.js:1774]:
```suggestion
            }
```

File: plugins/aws/sns/snsAppropriateSubscribers.js [hunk 4]
```diff
@@ -0,0 +1,68 @@
+var async = require('async');
+var helpers = require('../../../helpers/aws');
+
+module.exports = {
+    title: 'SNS Appropriate Subscribers',
+    category: 'SNS',
+    domain: 'Application Integration',
+    description: 'Ensure that all your Amazon SNS subscriptions are inspected for appropriate subscribers',
```
Comment (AkhtarAmir) [inline plugins/aws/sns/snsAppropriateSubscribers.js:8]:
```suggestion
    description: 'Ensure that Amazon SNS subscriptions are valid and there are no unwanted subscribers.',
```

File: plugins/aws/sns/snsAppropriateSubscribers.js [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    more_info: 'Amazon Simple Notification Service (Amazon SNS) is a managed service that provides message delivery from publishers to subscribers. So check for appropriate subsribers in order to improve access security to your SNS topics. ',
+    recommended_action: 'Ensure that no unwanted subscribers are using the subscriptions periodically.',
```
Comment (AkhtarAmir) [inline plugins/aws/sns/snsAppropriateSubscribers.js:10]:
```suggestion
    recommended_action: 'Check for unwanted SNS subscriptions periodically',
```

File: plugins/aws/sns/snsAppropriateSubscribers.js [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+    run: function(cache, settings, callback) {
+        var results = [];
+        var source = {};
+        var regions = helpers.regions(settings);
+
+        var config = {
+            sns_unwanted_subscribers_endpoint: settings.sns_unwanted_subscribers_endpoint || this.settings.sns_unwanted_subscribers_endpoint.default
+        };
+
```
Comment (AkhtarAmir) [inline plugins/aws/sns/snsAppropriateSubscribers.js:30]: Make this opt-in plugin

File: plugins/aws/sns/snsAppropriateSubscribers.js [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+                helpers.addResult(
+                    results, 0, 'No SNS subscriptions Found', region);
+                return rcb();
+            }
+
+            for (let i in listSubscriptions.data) {
+                let subscriber = listSubscriptions.data[i];
+                let resource = subscriber.SubscriptionArn;
+
+                if (subscriber.Endpoint == config.sns_unwanted_subscribers_endpoint){
```
Comment (AkhtarAmir) [inline plugins/aws/sns/snsAppropriateSubscribers.js:54]: Check if config includes endpoint

File: plugins/aws/sns/snsAppropriateSubscribers.js [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+    domain: 'Application Integration',
+    description: 'Ensure that all your Amazon SNS subscriptions are inspected for appropriate subscribers',
+    more_info: 'Amazon Simple Notification Service (Amazon SNS) is a managed service that provides message delivery from publishers to subscribers. So check for appropriate subsribers in order to improve access security to your SNS topics. ',
+    recommended_action: 'Ensure that no unwanted subscribers are using the subscriptions periodically.',
+    link: 'https://docs.aws.amazon.com/sns/latest/dg/sns-create-subscribe-endpoint-to-topic.html',
+    apis: ['SNS:listSubscriptions'],
+    settings: {
+        sns_unwanted_subscribers_endpoint: {
+            name: 'SNS Unwanted Subscribers Endpoint',
+            description: 'endpoint for SNS subscrptions that are unwanted',
```
Comment (AkhtarAmir) [inline plugins/aws/sns/snsAppropriateSubscribers.js:16]:
```suggestion
            description: 'Comma-separated list of subscription endpoint i.e. xyz@aquasec.com',
```

File: plugins/aws/sns/snsAppropriateSubscribers.js [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+                return rcb();
+            }
+
+            for (let i in listSubscriptions.data) {
+                let subscriber = listSubscriptions.data[i];
+                let resource = subscriber.SubscriptionArn;
+
+                if (subscriber.Endpoint == config.sns_unwanted_subscribers_endpoint){
+                    helpers.addResult(results, 2,
+                        'SNS subscriber is unwanted for topic', region, resource);
```
Comment (AkhtarAmir) [inline plugins/aws/sns/snsAppropriateSubscribers.js:56]:
```suggestion
                        'SNS subscription is an unwanted subscription', region, resource);
```

File: plugins/aws/sns/snsAppropriateSubscribers.js [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+                } else {
+                    helpers.addResult(results, 0,
+                        'SNS subscriber is appropriate for topic', region, resource);
```
Comment (AkhtarAmir) [inline plugins/aws/sns/snsAppropriateSubscribers.js:59]:
```suggestion
                        'SNS subscription is not an unwanted subscription', region, resource);
```

**Review by AkhtarAmir — CHANGES_REQUESTED** [review]

File: plugins/aws/sns/snsAppropriateSubscribers.js [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+        sns_unwanted_subscribers_endpoint: {
+            name: 'SNS Unwanted Subscribers Endpoint',
+            description: 'Comma-separated list of subscription endpoint i.e. xyz@aquasec.com',
+            regex: '^.*$',
+            default: 'sadeed1999@gmail.com',
+        }
+    },
+
+    run: function(cache, settings, callback) {
+        console.log(JSON.stringify(cache, null, 2))
```
Comment (sadeed12345) [inline plugins/aws/sns/snsAppropriateSubscribers.js:23]:
```suggestion
```

File: plugins/aws/sns/snsAppropriateSubscribers.js [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+            }
+
+            for (let i in listSubscriptions.data) {
+                let subscriber = listSubscriptions.data[i];
+                let resource = subscriber.SubscriptionArn;
+                let endpoint ;
+
+                if (subscriber.Endpoint){
+                    endpoint = subscriber.Endpoint.toLowerCase();
+                }
```
Comment (AkhtarAmir) [inline plugins/aws/sns/snsAppropriateSubscribers.js:61]:
```suggestion
```

File: plugins/aws/sns/snsAppropriateSubscribers.js [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+
+                if ((config.sns_unwanted_subscribers_endpoint.toLowerCase()).includes(endpoint)){
```
Comment (AkhtarAmir) [inline plugins/aws/sns/snsAppropriateSubscribers.js:63]:
```suggestion
                if (subscriber.Endpoint && config.sns_unwanted_subscribers_endpoint.toLowerCase().includes(subscriber.Endpoint.toLowerCase())){
```

File: plugins/aws/sns/snsAppropriateSubscribers.js [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+    more_info: 'Amazon Simple Notification Service (Amazon SNS) is a managed service that provides message delivery from publishers to subscribers. So check for appropriate subsribers in order to improve access security to your SNS topics. ',
+    recommended_action: 'Check for unwanted SNS subscriptions periodically',
+    link: 'https://docs.aws.amazon.com/sns/latest/dg/sns-create-subscribe-endpoint-to-topic.html',
+    apis: ['SNS:listSubscriptions'],
+    settings: {
+        sns_unwanted_subscribers_endpoint: {
+            name: 'SNS Unwanted Subscribers Endpoint',
+            description: 'Comma-separated list of subscription endpoint i.e. xyz@aquasec.com',
+            regex: '^.*$',
+            default: 'sadeed1999@gmail.com',
```
Comment (AkhtarAmir) [inline plugins/aws/sns/snsAppropriateSubscribers.js:18]:
```suggestion
            default: '',
```

**Review by AkhtarAmir — CHANGES_REQUESTED** [review]

**Review by AkhtarAmir — APPROVED** [review]

File: plugins/aws/sns/snsValidSubscribers.spec.js [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+    };
+};
+
+
+
+describe('snsValidSubscribers', function () {
+    describe('run', function () {
+        it('should PASS if SNS subscriber is appropriate for topic', function (done) {
+            const cache = createCache([listSubscriptions[0]]);
+            snsValidSubscribers.run(cache, { sns_unwanted_subscribers: 'sadeed1999@gmail.com' }, (err, results) => {
```
Comment (AkhtarAmir) [inline plugins/aws/sns/snsValidSubscribers.spec.js:53]:
```suggestion
        it('should PASS if SNS subscription is wanted', function (done) {
            const cache = createCache([listSubscriptions[0]]);
            snsValidSubscribers.run(cache, { sns_unwanted_subscribers: 'xyz@aquasec.com' }, (err, results) => {
```

File: plugins/aws/sns/snsValidSubscribers.spec.js [hunk 16]
```diff
@@ -0,0 +1,91 @@
...
+                expect(results.length).to.equal(1);
+                expect(results[0].status).to.equal(0);
+                expect(results[0].message).to.include('SNS subscription is not an unwanted subscription');
+                done();
+            });
+        });
+
+        it('should FAIL if SNS subscriber is unwanted for topic', function (done) {
```
Comment (AkhtarAmir) [inline plugins/aws/sns/snsValidSubscribers.spec.js:61]:
```suggestion
        it('should FAIL if SNS subscription is unwanted', function (done) {
```

File: plugins/aws/sns/snsValidSubscribers.spec.js [hunk 17]
```diff
@@ -0,0 +1,91 @@
...
+            const cache = createCache([listSubscriptions[1]]);
+            snsValidSubscribers.run(cache, { sns_unwanted_subscribers: 'sadeed1999@gmail.com' }, (err, results) => {
```
Comment (AkhtarAmir) [inline plugins/aws/sns/snsValidSubscribers.spec.js:63]:
```suggestion
            snsValidSubscribers.run(cache, { sns_unwanted_subscribers: 'xyz@aquasec.com' }, (err, results) => {
```

File: plugins/aws/sns/snsValidSubscribers.spec.js [hunk 18]
```diff
@@ -0,0 +1,91 @@
...
+                expect(results.length).to.equal(1);
+                expect(results[0].status).to.equal(2);
+                expect(results[0].message).to.include('SNS subscription is an unwanted subscription');
+                done();
+            });
+        });
+
+        it('should PASS if no SNS subscriptions found', function (done) {
+            const cache = createCache([]);
+            snsValidSubscribers.run(cache, { sns_unwanted_subscribers: 'sadeed1999@gmail.com' }, (err, results) => {
```
Comment (AkhtarAmir) [inline plugins/aws/sns/snsValidSubscribers.spec.js:73]:
```suggestion
            snsValidSubscribers.run(cache, { sns_unwanted_subscribers: 'xyz@aquasec.com' }, (err, results) => {
```

File: plugins/aws/sns/snsValidSubscribers.spec.js [hunk 19]
```diff
@@ -0,0 +1,91 @@
...
+                expect(results.length).to.equal(1);
+                expect(results[0].status).to.equal(0);
+                expect(results[0].message).to.include('No SNS subscriptions Found');
+                done();
+            });
+        });
+
+        it('should UNKNOWN if Unable to query for SNS subscriptions', function (done) {
+            const cache = createErrorCache();
+            snsValidSubscribers.run(cache, { sns_unwanted_subscribers: 'sadeed1999@gmail.com' }, (err, results) => {
```
Comment (AkhtarAmir) [inline plugins/aws/sns/snsValidSubscribers.spec.js:83]:
```suggestion
            snsValidSubscribers.run(cache, { sns_unwanted_subscribers: 'xyz@aquasec.com' }, (err, results) => {
```

File: plugins/aws/sns/snsValidSubscribers.js [hunk 20]
```diff
@@ -0,0 +1,74 @@
...
+            for (let i in listSubscriptions.data) {
+                let subscriber = listSubscriptions.data[i];
+                let resource = subscriber.SubscriptionArn;
+
+                if (subscriber.Endpoint && config.sns_unwanted_subscribers.includes(subscriber.Endpoint.toLowerCase())){
+                    helpers.addResult(results, 2,
+                        'SNS subscription is an unwanted subscription', region, resource);
+                } else {
+                    helpers.addResult(results, 0,
+                        'SNS subscription is not an unwanted subscription', region, resource);
```
Comment (AkhtarAmir) [inline plugins/aws/sns/snsValidSubscribers.js:65]:
```suggestion
                        'SNS subscription is a wanted subscription', region, resource);
```

File: plugins/aws/sns/snsValidSubscribers.spec.js [hunk 21]
```diff
@@ -0,0 +1,91 @@
...
+
+
+describe('snsValidSubscribers', function () {
+    describe('run', function () {
+        it('should PASS if SNS subscription is wanted', function (done) {
+            const cache = createCache([listSubscriptions[0]]);
+            snsValidSubscribers.run(cache, { sns_unwanted_subscribers: 'xyz@aquasec.com' }, (err, results) => {
+                expect(results.length).to.equal(1);
+                expect(results[0].status).to.equal(0);
+                expect(results[0].message).to.include('SNS subscription is not an unwanted subscription');
```
Comment (AkhtarAmir) [inline plugins/aws/sns/snsValidSubscribers.spec.js:56]:
```suggestion
                expect(results[0].message).to.include('SNS subscription is a wanted subscription');
```

File: plugins/aws/sns/snsValidSubscribers.spec.js [hunk 22]
```diff
@@ -0,0 +1,91 @@
+var expect = require('chai').expect;
+const snsValidSubscribers = require('./snsValidSubscribers');
+
+const listSubscriptions = [
+    {
+        "SubscriptionArn": "arn:aws:sns:us-east-1:000011112222:Default_CloudWatch_Alarms_Topic:18c0d0b0-f39e-4ac8-8e08-88f239768f61",
+        "Owner": "000011112222",
+        "Protocol": "email",
+        "Endpoint": "makhtar.pucit@gmail.com",
```
Comment (AkhtarAmir) [inline plugins/aws/sns/snsValidSubscribers.spec.js:9]:
```suggestion
        "Endpoint": "xyz@aquasec.com",
```

File: plugins/aws/sns/snsValidSubscribers.spec.js [hunk 23]
```diff
@@ -0,0 +1,91 @@
...
+        "TopicArn": "arn:aws:sns:us-east-1:000011112222:Default_CloudWatch_Alarms_Topic"
+    },
+    {
+        "SubscriptionArn": "arn:aws:sns:us-east-1:000011112222:aqua-cspm-sns-000011112222:1a1450c0-03bd-4feb-a99f-bba3ca0540e9",
+        "Owner": "000011112222",
+        "Protocol": "email",
+        "Endpoint": "sadeed1999@gmail.com",
```
Comment (AkhtarAmir) [inline plugins/aws/sns/snsValidSubscribers.spec.js:16]:
```suggestion
        "Endpoint": "xxz@aquasec.com",
```

File: plugins/aws/sns/snsValidSubscribers.spec.js [hunk 24]
```diff
@@ -0,0 +1,91 @@
...
+        }
+    };
+};
+
+
+
+describe('snsValidSubscribers', function () {
+    describe('run', function () {
+        it('should PASS if SNS subscription is wanted', function (done) {
+            const cache = createCache([listSubscriptions[0]]);
```
Comment (AkhtarAmir) [inline plugins/aws/sns/snsValidSubscribers.spec.js:52]:
```suggestion
            const cache = createCache([listSubscriptions[1]]);
```

File: plugins/aws/sns/snsValidSubscribers.spec.js [hunk 25]
```diff
@@ -0,0 +1,91 @@
...
+            snsValidSubscribers.run(cache, { sns_unwanted_subscribers: 'xyz@aquasec.com' }, (err, results) => {
+                expect(results.length).to.equal(1);
+                expect(results[0].status).to.equal(0);
+                expect(results[0].message).to.include('SNS subscription is a wanted subscription');
+                done();
+            });
+        });
+
+        it('should FAIL if SNS subscription is unwanted', function (done) {
+            const cache = createCache([listSubscriptions[1]]);
```
Comment (AkhtarAmir) [inline plugins/aws/sns/snsValidSubscribers.spec.js:62]:
```suggestion
            const cache = createCache([listSubscriptions[2]]);
```

File: plugins/aws/sns/snsValidSubscribers.js [hunk 26]
```diff
@@ -0,0 +1,74 @@
...
+
+            if (!listSubscriptions.data || !listSubscriptions.data.length) {
+                helpers.addResult(
+                    results, 0, 'No SNS subscriptions Found', region);
+                return rcb();
+            }
+
+            for (let i in listSubscriptions.data) {
+                let subscriber = listSubscriptions.data[i];
+                let resource = subscriber.SubscriptionArn;
```
Comment (AkhtarAmir) [inline plugins/aws/sns/snsValidSubscribers.js:58]:
```suggestion
            for (let subscriber of listSubscriptions.data) {
                if (!subscriber.SubscriptionArn) continue;

                let resource = subscriber.SubscriptionArn;
```

File: plugins/aws/sns/snsValidSubscribers.spec.js [hunk 27]
```diff
@@ -0,0 +1,91 @@
...
+            snsValidSubscribers.run(cache, { sns_unwanted_subscribers: 'xyz@aquasec.com' }, (err, results) => {
+                expect(results.length).to.equal(1);
+                expect(results[0].status).to.equal(0);
+                expect(results[0].message).to.include('SNS subscription is a wanted subscription');
+                done();
+            });
+        });
+
+        it('should FAIL if SNS subscription is unwanted', function (done) {
+            const cache = createCache([listSubscriptions[2]]);
```
Comment (AkhtarAmir) [inline plugins/aws/sns/snsValidSubscribers.spec.js:62]:
```suggestion
            const cache = createCache([listSubscriptions[0]]);
```

**Pull request merged**